src/neural_de.jl

```diff
@@ -502,3 +502,54 @@ augment(x::AbstractArray{S, T}, augment_dim::Int) where {S, T} =
 
 Base.getproperty(ande::AugmentedNDELayer, sym::Symbol) =
     hasproperty(ande, sym) ? getfield(ande, sym) : getfield(ande.nde, sym)
+
+
+struct NeuralHamiltonianDE{M,P,RE,T,A,K} <: NeuralDELayer
+    model::M
+    p::P
+    re::RE
+    tspan::T
+    args::A
+    kwargs::K
+
+    function NeuralHamiltonianDE(model,tspan,args...;p = nothing,kwargs...)
+        _p,re = Flux.destructure(model)
+        if p === nothing
+            p = _p
+        end
+        new{typeof(model),typeof(p),typeof(re),
+            typeof(tspan),typeof(args),typeof(kwargs)}(
+            model,p,re,tspan,args,kwargs)
+    end
+
+    function NeuralHamiltonianDE(model::FastChain,tspan,args...;p = initial_params(model),kwargs...)
+        re = nothing
+        new{typeof(model),typeof(p),typeof(re),
+            typeof(tspan),typeof(args),typeof(kwargs)}(
+            model,p,re,tspan,args,kwargs)
+    end
+end
+
+function (n::NeuralHamiltonianDE)(x,p=n.p)
+    function neural_hamiltonian!(du,u,p,t)
+        H = Flux.gradient(u -> n.re(p)(u)[1], u)[1]
+        n = size(x)[1] ÷ 2
+        du[1:n] .= -H[1:n]
+        du[(n + 1):end] .= H[(n + 1):end]
```

Comment:
I am not sure how to do it in a way that is GPU friendly. Any suggestions for improvement?

Reply:
using Slices like that is okay, it's just two kernels. But you can make it one kernel by using a fill array with a `-1` and a 1, or just allocating the array of 1 and -1 (maybe caching it). That should be a relatively minor improvement though.